draft-irtf-cfrg-opaque.md

```diff
@@ -959,11 +990,103 @@ The protocol runs as shown below:
+
+- password, the input password; and
+- blind, the random blinding scalar returned by `Blind()`, of length Nok; and
+- client_ake_state, the client's AKE state if necessary.
+
+The server state may have the following fields:
+
+- server_ake_state, the server's AKE state if necessary.
+
 The rest of this section describes these authenticated key exchange messages
 and their parameters in more detail. {{cred-retrieval}} discusses internal
 functions used for retrieving client credentials, and {{ake-protocol}} discusses
 how these functions are used to execute the authenticated key exchange protocol.
 
+## External Client API {#opaque-client}
+
+~~~
+ClientInit(password)
+
+State:
+- state, a ClientState structure.
+
+Input:
+- password, an opaque byte string containing the client's password.
+
+Output:
+- ke1, a KE1 message structure.
+
+Steps:
+1. request, blind = CreateCredentialRequest(password)
+2. state.blind = blind
+3. ake_1 = Start(request)
+4. Output KE1(request, ake_1)
+~~~
+
+~~~
+ClientFinish(client_identity, server_identity, ke2)
+
+State:
+- state, a ClientState structure
+
+Input:
+- client_identity, the optional encoded client identity, which is set
+  to client_public_key if not specified.
+- server_identity, the optional encoded server identity, which is set
+  to server_public_key if not specified.
+- ke2, a KE2 message structure.
+
+Output:
+- ke3, a KE3 message structure.
+- session_key, the session's shared secret.
+- export_key, an additional client key.
+
+Steps:
+1. (client_private_key, server_public_key, export_key) =
+    RecoverCredentials(state.password, state.blind, ke2.CredentialResponse,
+                       server_identity, client_identity)
+2. (ke3, session_key) =
+    ClientFinalize(client_identity, client_private_key, server_identity,
+                    server_public_key, ke2)
+3. Output (ke3, session_key)
+~~~
+
+## External Server API {#opaque-server}
+
+~~~
+ServerInit(server_identity, server_private_key, server_public_key,
+           record, credential_identifier, oprf_seed, ke1)
+
+Input:
+- server_identity, the optional encoded server identity, which is set to
+  server_public_key if nil.
+- server_private_key, the server's private key.
+- server_public_key, the server's public key.
+- record, the client's RegistrationRecord structure.
+- credential_identifier, an identifier that uniquely represents the credential.
+- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key.
+- ke1, a KE1 message structure.
+
+Output:
+- ke2, a KE2 structure.
+
+Steps:
+1. response = CreateCredentialResponse(ke1.request, server_public_key, record,
+    credential_identifier, oprf_seed)
+2. ake_2 = Response(server_identity, server_private_key,
+    client_identity, record.client_public_key, ke1, response)
+3. Output KE2(response, ake_2)
+~~~
+
+Since the OPRF is a two-message protocol, KE3 has no element of the OPRF. We can
+therefore call the AKE's `ServerFinish()` directly and return its output.
```

Comment:
Don't we still need a function to call to get the `session_key` on the server side? (ServerFinish is the function for this, but I don't see why it's defined way below and not here?)

Reply:
Yeah, but I was hesitant in describing it here, since this part is generic without any details about the AKE. Specifying ServerFinish here will start talking about 3DH before the dedicated AKE section (if that makes sense)

Reply:
Well, wouldn't _any_ AKE need this final function on the server? How does the server get the session key otherwise?

Reply:
Of course, but the internals of this function differ from one AKE to the other

Reply:
You're right, the function should still be described here. I described the necessary behaviour and API constraints of ServerFinish() in general. What do you think of that update?